pkg/userhelpers/user_account_utils.go

```diff
@@ -21,59 +26,127 @@ type OktaAccountInfo struct {
+type GetUserInfoFunc func(ctx context.Context, username string) (*models.UserInfo, error)
+
 // GetOrCreateUserAccount will return an account if it exists, or create and return a new one if not
-func GetOrCreateUserAccount(store *storage.Store, username string, useLocal bool, baseURL string, token string, isMacUser bool) (*authentication.UserAccount, error) {
+func GetOrCreateUserAccount(ctx context.Context, store *storage.Store, username string, hasLoggedIn bool,
+	isMacUser bool, getAccountInformation GetAccountInfoFunc) (*authentication.UserAccount, error) {
 	userAccount, accErr := store.UserAccountGetByUsername(username)
 	if accErr != nil {
 		return nil, errors.New("failed to get user information from the database")
 	}
-	if userAccount != nil {
+	if userAccount != nil && userAccount.HasLoggedIn {
 		return userAccount, nil
 	}
-	accountInfo := &OktaAccountInfo{}
-	if useLocal {
-		accountInfo.Name = username + " Doe"
-		accountInfo.Locale = "en_US"
-		accountInfo.Email = username + "@local.cms.gov"
-		accountInfo.PreferredUsername = username
-		accountInfo.GivenName = username
-		accountInfo.FamilyName = "Doe"
-		accountInfo.ZoneInfo = "America/Los_Angeles"
-	} else {
-		oktaInfo, err := GetUserInfoFromOkta(baseURL, token)
-		if err != nil {
-			return nil, err
-		}
-		accountInfo = oktaInfo
+	accountInfo, err := getAccountInformation(ctx, username)
+	if err != nil {
+		return nil, err
 	}
 
 	if userAccount == nil {
-		user := authentication.UserAccount{
-			Username:   &username,
-			IsEUAID:    !isMacUser,
-			CommonName: accountInfo.Name,
-			Locale:     accountInfo.Locale,
-			Email:      accountInfo.Email,
-			GivenName:  accountInfo.GivenName,
-			FamilyName: accountInfo.FamilyName,
-			ZoneInfo:   accountInfo.ZoneInfo,
-		}
-		newAccount, err := store.UserAccountInsertByUsername(&user)
-		if err != nil {
-			return nil, err
+		userAccount = &authentication.UserAccount{}
+	}
+	userAccount.Username = &username
+	userAccount.IsEUAID = !isMacUser
+	userAccount.CommonName = accountInfo.Name
+	userAccount.Locale = accountInfo.Locale
+	userAccount.Email = accountInfo.Email
+	userAccount.GivenName = accountInfo.GivenName
+	userAccount.FamilyName = accountInfo.FamilyName
+	userAccount.ZoneInfo = accountInfo.ZoneInfo
+	userAccount.HasLoggedIn = hasLoggedIn
+
+	if userAccount.ID == uuid.Nil {
+		newAccount, newErr := store.UserAccountInsertByUsername(userAccount)
+		if newErr != nil {
+			return nil, newErr
 		}
-		userAccount = newAccount
+		return newAccount, nil
 	}
-	return userAccount, nil
+
+	updatedAccount, updateErr := store.UserAccountUpdateByUserName(userAccount)
+	if updateErr != nil {
+		return nil, updateErr
+	}
+	return updatedAccount, nil
+}
+
+// GetUserInfoAccountInformationWrapperFunction returns a function that returns *AccountInfo with the input of a function that returns UserInfo
+func GetUserInfoAccountInformationWrapperFunction(getAccountInformation GetUserInfoFunc) GetAccountInfoFunc {
```

Comment:
Could we rename these functions with `AccountInformation` to `GetUserInfoAccountInfoWrapperFunction` or even `GetUserInfoAccountInfoWrapperFunc` to keep the naming consistent or was there a reason you put `AccountInformation`?

Reply:
Thanks Riley! Good suggestions, I'll change it.